superset-frontend/plugins/plugin-chart-table/src/controlPanel.tsx

```diff
@@ -517,6 +517,11 @@ const config: ControlPanelConfig = {
       ],
     },
   ],
+  denormalizeFormData: formData => ({
+    ...formData,
+    metrics: formData.standardized_form_data.sharedFormData.metrics,
```

Comment:
I'm wondering if we should consider adding `percent_metrics` to the metrics array in `sharedFormData`? Let's say we have a table chart with just a single metric in `percentage_metrics` and change to Pie - currently the metric will be lost.

As a solution - Maybe we could add a property to the normalized control values that states some additional context about the origin of the value. In this case it could specify that it originated from the `percent_metrics` control rather than a "regular" metric control. So the additional value coult, for instance, be `sourceControl`.

This would also address the problem identified in the Pivot table - if we had the additional context about what type of a control the value originated from, we would more easily be able to map them back to their original controls (`groupbyRows` or `groupbyColums`). Ping @kgabryje

Reply:
Not sure if I follow - it looks to me like rows and columns in pivot table do return to their original states?

https://user-images.githubusercontent.com/15073128/168563992-9d9749fb-61e4-4664-b0a5-f663a4ac40df.mov

Or do you mean that, for example, if both columns and rows in pivot table contain a value `name`, then it gets deduplicated when switching to table but `sourceControl` contains information that `name` originates from both `columns` and `rows` in pivot table? If that's the point, I like this idea!

Reply:
Thanks for the response - I must have had something wrong with my devenv, as after I reloaded it worked just like you described 👍 To add to this comment:

> So I think that when we transition from Pivot Table to any other chart we should join values from Rows and Columns, remove duplicates and populate Dimensions of the new viz with the result

I think we should keep the duplicates in the normalized form data, but we could enrich the interface so that we know where the values came from:

```typescript
export interface SharedFormData {
  metrics: {
    value: QueryFormMetric;
    sourceViz: string;
    sourceControl: string;
  }[];
  columns: {
    value: QueryFormColumn;
    sourceViz: string;
    sourceControl: string;
  }[];
}
```

This way the pie chart might not care if the metric was from `metrics` or `percent_metrics`, but some other chart might decide to map the values over differently, depending on where the value came from. This is especially true for the `x_axis` control, which in a sense is a column/groupby, but should be handled differently from the series control.

Reply:
> I think we should keep the duplicates in the normalized form data

I'm still not really sold on this, because we can enter a state which is not allowed via user interface

Reply:
> I'm wondering if we should consider adding `percent_metrics` to the metrics array in `sharedFormData`? Let's say we have a table chart with just a single metric in `percentage_metrics` and change to Pie - currently the metric will be lost.

This is an interesting question. It is also a legacy of Superset's history. The `Percentage Metrics` is a post-processing **metric** rather than a SQL base **metric**. basically, it is not simply mapping from a source viz to a target viz. The same control is `Contribution Mode` in the Line chart V2.

I have said before, to solve this issue, I think we eventually need to move all post-processing(AA) into the metric and column so that we can do the real map between different charts.

Reply:
>This way the pie chart might not care if the metric was from metrics or percent_metrics, but some other chart might decide to map the values over differently, depending on where the value came from. This is especially true for the x_axis control, which in a sense is a column/groupby, but should be handled differently from the series control.

The `xaxis` didn't put in the `sharedControls`. it's a publicControls. In the other words, the x-axis just "inhert" latest chart x-axis value, it can not mapping from `sharedControls`.

I suggest that the modeling(like metric/dimension/filter/etc...) should decouple from visualization. The `sourceControl: string;` means that preserve the control context, this will allow modeling and visualization to be mixed.